src/test/util_string_tests.cpp

```diff
@@ -13,17 +13,57 @@ BOOST_AUTO_TEST_SUITE(util_string_tests)
 // Helper to allow compile-time sanity checks while providing the number of
 // args directly. Normally PassFmt<sizeof...(Args)> would be used.
 template <unsigned NumArgs>
-inline void PassFmt(util::ConstevalFormatString<NumArgs> fmt)
+inline void PassFmt(ConstevalFormatString<NumArgs> fmt)
 {
-    // This was already executed at compile-time, but is executed again at run-time to avoid -Wunused.
+    // Exercise happy paths at run-time for code coverage metrics.
     decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
+
+    // Prove parity with tinyformat
+    switch (NumArgs) {
```

Comment:
I like the idea of testing the validator together with the implementation!

But I'm not in love with the current approach.

Why are we testing `PassFmt<1>("%02d");` with `foo` - I don't think it helps with understanding how formatter works.

Also `PassFmt<1>("%s")` made sense when we were only validating the number of args, but we've extended it since, I think we should extend examples with the actual parameters, e.g.
* `PassFmt("%s", "test");`
* `PassFmt("%02d", 42);`
* `PassFmt("%12$s %2$s %1$s", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");`

In which case each example would document the format that we're exercising more specifically.

It would also obviate the parameter count by examples, would all be runtime (seems to me that would simplify a lot) and comparison would simply depend on the behavior of `tfm::format`, e.g. validation would be something like:
```C++
template <typename... Args>
inline bool ValidateFmt(const char* fmt, const char* expected_error, const Args&... args)
{
    try {
        tfm::format(fmt, args...);
        ConstevalFormatString<sizeof...(Args)>::Detail_CheckNumFormatSpecifiers(fmt);
        return true;
    } catch (const tfm::format_error&) {
        using ErrType = const char*;
        auto check_throw = [expected_error](const ErrType& str) { return str == expected_error; };
        BOOST_CHECK_EXCEPTION(ConstevalFormatString<sizeof...(Args)>::Detail_CheckNumFormatSpecifiers(fmt), ErrType, check_throw);
        return false;
    }
}

template <typename... Args>
void PassFmt(const char* fmt, const Args&... args) { BOOST_CHECK(ValidateFmt(fmt, nullptr, args...)); }

template <typename... Args>
void FailFmt(const char* fmt, const char* expected_error, const Args&... args) { BOOST_CHECK(!ValidateFmt(fmt, expected_error, args...)); }
```

and usage would be something like
```
    PassFmt("%%");
    PassFmt("%s", "test");

    auto err_num{"Format specifier count must match the argument count!"};
    FailFmt("%s", err_num);
    FailFmt("%s", err_num, "test", "extra");
```

Reply:
I like parts of your direction here, and have taken the time to try it out in 2fb2e72f45a1c81d036515861713371fec2bbe2c. My aim is to retain the tests of leaving out one arg though (which your suggestion doesn't include). I'm not enough of a variadic template arg magician to figure out how to skip the last argument, which forces me to still lean on maflcko's `tuple_cat` approach, leaving it feeling somewhat half-baked.

Reply:
> My aim is to retain the tests of leaving out one arg though (which your suggestion doesn't include)

I didn't include it since I though a single instance of those is enough in the tests.

I would prefer having concrete typed examples over retesting -1 and +1 args (for which explicit examples should likely suffice)

Reply:
I prefer retaining negative tests to increase certainty and would rather not do the `PassFmt<1>("%s")` -> `PassFmt("%s", "test")` refactor in this PR if that's okay with you.